src/types/mutex.rs

```diff
@@ -0,0 +1,402 @@
+        }
+    }
+
+    #[inline]
+    #[cfg(not(panic = "unwind"))]
+    pub fn done(&self, _guard: &Guard) {}
+
+    #[inline]
+    #[cfg(panic = "unwind")]
+    fn get(&self) -> bool {
+        self.failed.load(Ordering::Relaxed)
+    }
+
+    #[inline(always)]
+    #[cfg(not(panic = "unwind"))]
+    pub fn get(&self) -> bool {
+        false
+    }
+
+    #[inline]
+    fn clear(&self) {
+        #[cfg(panic = "unwind")]
+        self.failed.store(false, Ordering::Relaxed)
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct Guard {
+    #[cfg(panic = "unwind")]
+    panicking: bool,
+}
+
+/// Wrapper for [`PyMutex`](https://docs.python.org/3/c-api/init.html#c.PyMutex), exposing an RAII guard interface.
+///
+/// Compared with `std::sync::Mutex` or `parking_lot::Mutex`, this is a very
+/// stripped-down locking primitive that only supports blocking lock and unlock
+/// operations and does not support `try_lock` or APIs that depend on
+/// `try_lock`.  For this reason, it is not possible to avoid the possibility of
+/// possibly blocking when calling `lock` and extreme care must be taken to avoid
+/// introducing a deadlock.
+///
+/// This type is most useful when arbitrary Python code might execute while the
+/// lock is held. On the GIL-enabled build, PyMutex will release the GIL if the
+/// thread is blocked on acquiring the lock. On the free-threaded build, threads
+/// blocked on acquiring a PyMutex will not prevent the garbage collector from
+/// running.
+///
+/// ## Poisoning
+///
+/// Like `std::sync::Mutex`, `PyMutex` implements poisoning. A mutex
+/// is considered poisoned whenever a thread panics while holding the mutex. Once
+/// a mutex is poisoned, all other threads are unable to access the data by
+/// default as it is likely to be tainted (some invariant is not being held).
+///
+/// This means that the `lock` method returns a `Result` which indicated whether
+/// the mutex has been poisoned or not. Must usage will simple `unwrap()` these
+/// results, propagating panics among threads to ensure a possible invalid
+/// invariant is not being observed.
+///
+/// A poisoned mutex, however, does not prevent all access to the underlying
+/// data. The `PoisonError` type has an `into_inner` method which will return
+/// the guard that would have otherwise been returned on a successful lock. This
+/// allows access to the data, despite the lock being poisoned.
+pub struct PyMutex<T: ?Sized> {
+    mutex: UnsafeCell<crate::ffi::PyMutex>,
+    poison: Flag,
+    data: UnsafeCell<T>,
+}
+
+/// RAII guard to handle releasing a PyMutex lock.
+///
+/// The lock is released when `PyMutexGuard` is dropped.
+pub struct PyMutexGuard<'a, T: ?Sized> {
+    inner: &'a PyMutex<T>,
+    poison: Guard,
+}
+
+/// `T` must be `Sync` for a [`PyMutexGuard<T>`] to be `Sync`
+/// because it is possible to get a `&T` from `&MutexGuard` (via `Deref`).
+unsafe impl<T: ?Sized + Sync> Sync for PyMutexGuard<'_, T> {}
```

Comment:
We should probably add a test to assert that `PyMutexGuard` is not send - I think this would presumably be a problem as other threads should not be able to unlock a locked mutex?

#5308 can help with writing test for not send (just negate the example assertions there).